From 67df1daf887fad98f9c84e6304d725d3762beb99 Mon Sep 17 00:00:00 2001
From: Tao Huang <huangtao@rock-chips.com>
Date: Sat, 28 Mar 2020 16:43:53 +0800
Subject: [PATCH] rfkill: Introduce Rockchip RFKILL driver

Allow rfkill-bt.c and rfkill-wlan.c build as module.
Add CONFIG_RFKILL_RK which default y.

Change-Id: I1d399641587124d2291b5b49e5538e7e490a843a
Signed-off-by: Tao Huang <huangtao@rock-chips.com>
---
 include/linux/rfkill-wlan.h | 3 +++
 net/rfkill/Kconfig          | 8 ++++++++
 net/rfkill/Makefile         | 3 ++-
 net/rfkill/rfkill-bt.c      | 6 ++++++
 net/rfkill/rfkill-wlan.c    | 7 ++-----
 5 files changed, 21 insertions(+), 6 deletions(-)

diff --git a/include/linux/rfkill-wlan.h b/include/linux/rfkill-wlan.h
index d915e63fb6b8..5e00f8808e6f 100644
--- a/include/linux/rfkill-wlan.h
+++ b/include/linux/rfkill-wlan.h
@@ -56,4 +56,7 @@ int rockchip_wifi_get_oob_irq_flag(void);
 int rockchip_wifi_reset(int on);
 int rockchip_wifi_mac_addr(unsigned char *buf);
 void *rockchip_wifi_country_code(char *ccode);
+int rfkill_wlan_init(void);
+void rfkill_wlan_exit(void);
+
 #endif
diff --git a/net/rfkill/Kconfig b/net/rfkill/Kconfig
index 060600b03fad..c40a4241ff03 100644
--- a/net/rfkill/Kconfig
+++ b/net/rfkill/Kconfig
@@ -31,3 +31,11 @@ config RFKILL_GPIO
 	help
 	  If you say yes here you get support of a generic gpio RFKILL
 	  driver.
+
+config RFKILL_RK
+	tristate "Rockchip RFKILL driver"
+	depends on RFKILL
+	depends on MMC
+	default y
+	help
+	  Rockchip rfkill driver for rk29/rk3X
diff --git a/net/rfkill/Makefile b/net/rfkill/Makefile
index ed13f8e7f6c7..6ce18d54a8d8 100644
--- a/net/rfkill/Makefile
+++ b/net/rfkill/Makefile
@@ -6,4 +6,5 @@ rfkill-y			+= core.o
 rfkill-$(CONFIG_RFKILL_INPUT)	+= input.o
 obj-$(CONFIG_RFKILL)		+= rfkill.o
 obj-$(CONFIG_RFKILL_GPIO)	+= rfkill-gpio.o
-obj-y                           += rfkill-wlan.o rfkill-bt.o
+rfkill-rk-y			+= rfkill-wlan.o rfkill-bt.o
+obj-$(CONFIG_RFKILL_RK)		+= rfkill-rk.o
diff --git a/net/rfkill/rfkill-bt.c b/net/rfkill/rfkill-bt.c
index 99acb9813eff..58ad5b073809 100644
--- a/net/rfkill/rfkill-bt.c
+++ b/net/rfkill/rfkill-bt.c
@@ -786,7 +786,12 @@ static struct platform_driver rfkill_rk_driver = {
 
 static int __init rfkill_rk_init(void)
 {
+	int err;
+
 	LOG("Enter %s\n", __func__);
+	err = rfkill_wlan_init();
+	if (err)
+		return err;
 	return platform_driver_register(&rfkill_rk_driver);
 }
 
@@ -794,6 +799,7 @@ static void __exit rfkill_rk_exit(void)
 {
 	LOG("Enter %s\n", __func__);
 	platform_driver_unregister(&rfkill_rk_driver);
+	rfkill_wlan_exit();
 }
 
 module_init(rfkill_rk_init);
diff --git a/net/rfkill/rfkill-wlan.c b/net/rfkill/rfkill-wlan.c
index 11a973e2bf53..27bf91f83f80 100644
--- a/net/rfkill/rfkill-wlan.c
+++ b/net/rfkill/rfkill-wlan.c
@@ -836,21 +836,18 @@ static struct platform_driver rfkill_wlan_driver = {
 	},
 };
 
-static int __init rfkill_wlan_init(void)
+int __init rfkill_wlan_init(void)
 {
 	LOG("Enter %s\n", __func__);
 	return platform_driver_register(&rfkill_wlan_driver);
 }
 
-static void __exit rfkill_wlan_exit(void)
+void __exit rfkill_wlan_exit(void)
 {
 	LOG("Enter %s\n", __func__);
 	platform_driver_unregister(&rfkill_wlan_driver);
 }
 
-module_init(rfkill_wlan_init);
-module_exit(rfkill_wlan_exit);
-
 MODULE_DESCRIPTION("rock-chips rfkill for wifi v0.1");
 MODULE_AUTHOR("gwl@rock-chips.com");
 MODULE_LICENSE("GPL");
-- 
2.35.3

